src/linalg.jl

```diff
@@ -629,6 +629,61 @@ function dot(A::AbstractSparseMatrixCSC, B::Union{DenseMatrixUnion,WrapperMatrix
     return conj(dot(B, A))
 end
 
+function dot(x::AbstractSparseVector, Q::Diagonal, y::AbstractVector)
+    if length(x) != length(y)
+        throw(
+            DimensionMismatch("Vector x has a length $(length(x)) but y has a length $(length(y))")
+        )
+    end
+    d = Q.diag
+    nzvals = nonzeros(x)
+    nzinds = nonzeroinds(x)
+    s = zero(Base.promote_eltype(x, Q, y))
```

Comment:
In `LinearAlgebra`, we use

```julia
s = zero(typeof(dot(first(x), first(A), first(y))))
```

because `dot` acts recursively and technically, arrays of arrays are allowed.

Reply:
done!